### ISSUE TITLE
discrepancy in sharding

### ISSUE THREAD

**Issue by alperyilmaz** [xiaodaigh/disk.frame#79]
This case is related to #77 where reading multiple csv files were discussed

When I use `shardby` I get discrepant results for counts. Below is the summary of my findings

| UniqueCarrier | disk.frame shard | disk.frame **not_sharded** | actual  |
|---------------|------------------|------------------------|---------|
| AA            | 5263987          | 6655735                | 6655735 |
| AS            | 795393           | 994764                 | 994764  |
| CO            | 3187682          | 4017862                | 4017862 |
| DL            | 6247363          | 7884309                | 7884309 |
| EA            | 701362           | 919785                 | 919785  |
| HP            | 1434364          | 1809250                | 1809250 |
| ML (1)        | 56231            | 70622                  | 70622   |
| NW            | 3502394          | 4410734                | 4410734 |
| PA (1)        | 240546           | 316167                 | 316167  |
| PI            | 672802           | 873957                 | 873957  |
| PS            | 70563            | 83617                  | 83617   |
| TW            | 1918283          | 2421955                | 2421955 |
| UA            | 4731504          | 5938506                | 5938506 |
| US            | 5752829          | 7295919                | 7295919 |
| WN            | 3428952          | 4231882                | 4231882 |

More detailed analysis for [sharded](http://rpubs.com/alperyilmaz/disk-frame-shard) and [not-sharded](http://rpubs.com/alperyilmaz/disk-frame-shard2) approaches are available.

**Comment by xiaodaigh** [comment]
This is a serious bug. Thanks for reporting. I will look into this ASAP.

**Comment by xiaodaigh** [comment]
```r
library(disk.frame)

pt = proc.time()
non_shard = 2007:2008 %>% 
  purrr::map(~ csv_to_disk.frame(paste0("inst/flights_case_study/data/",.x,".csv"), 
                                 outdir=paste0("tmp",.x),
                                 overwrite=T)) %>% 
  rbindlist.disk.frame(
    outdir = "inst/data/flights_all_no_shard.df", by_chunk_id = T)


sharded = 2007:2008 %>% 
  purrr::map(~ csv_to_disk.frame(paste0("inst/flights_case_study/data/",.x,".csv"), 
                                 outdir=paste0("tmp2",.x),
                                 shardby=c("Year","Month"),
                                 overwrite=T)) %>% 
  rbindlist.disk.frame(
    outdir = "inst/data/flights_all_shard.df", by_chunk_id = T)



nrow(non_shard)
nrow(sharded)

library(disk.frame)
sharded = disk.frame("inst/data/flights_all_shard.df")
a = sharded[,.N, UniqueCarrier, keep="UniqueCarrier"]

a = a[,sum(N), UniqueCarrier]

nosharded = disk.frame("inst/data/flights_all_no_shard.df/")
a1 = nosharded[,.N, UniqueCarrier, keep="UniqueCarrier"]
a1 = a1[,sum(N), UniqueCarrier]

setkey(a, UniqueCarrier)
setkey(a1, UniqueCarrier)

identical(a, a1)
```

the above seems fine. so the error could be from using in_chunk_size or loading more than two years. Will try to debug ASAP

**Comment by xiaodaigh** [comment]
Happy to report that this has been fixed by this commit (ea11162197c719ca3ac869cbf19dce67a2879cf8)

the bug was caused by gaps in the chunk ids which has been fixed in the development branch, but i've bought the fix into master now. So please reinstall `disk.frame` and test again

```r
remotes::install_github("xiaodaigh/disk.frame")
```

**Testing code**

```r
library(disk.frame)
plan(multiprocess)
library(magrittr)

a = disk.frame("flights_all_sharded.df")
b = disk.frame("flights_all_notsharded.df")

a2 = a %>% 
  group_by(UniqueCarrier) %>% 
  summarise(n=n()) %>% 
  collect %>% 
  group_by(UniqueCarrier) %>% 
  summarise(n=sum(n)) %>% 
  arrange(UniqueCarrier)

b2 = b %>% 
  group_by(UniqueCarrier) %>% 
  summarise(n=n()) %>% 
  collect %>% 
  group_by(UniqueCarrier) %>% 
  summarise(n=sum(n)) %>% 
  arrange(UniqueCarrier)

identical(a2, b2)
```
